### PR TITLE
Bug 1345391 - Support PaymentRequest::canMakePayment flow webAPI part

### PR DESCRIPTION
**Pull request by alphan102**

### CODE REVIEW

File: dom/payments/PaymentRequest.cpp [hunk 1]
```diff
@@ -148,12 +150,42 @@ PaymentRequest::Show(ErrorResult& aRv)
...
 PaymentRequest::CanMakePayment(ErrorResult& aRv)
 {
-  return nullptr;
+  if (mState != eCreated) {
+    aRv.Throw(NS_ERROR_DOM_INVALID_STATE_ERR);
+    return nullptr;
+  }
+
+  if (mResultPromise) {
+    aRv.Throw(NS_ERROR_DOM_QUOTA_EXCEEDED_ERR);
```
Comment (marcoscaceres) [inline dom/payments/PaymentRequest.cpp]: Oh, so this part of the spec is about to change:
https://github.com/w3c/browser-payment-api/pull/488

Comment (marcoscaceres) [inline dom/payments/PaymentRequest.cpp]: (waiting on OK from Apple that that is what they want)

Comment (alphan102) [inline dom/payments/PaymentRequest.cpp]: I will update the error with "NotAllowedError".

File: dom/payments/PaymentRequest.cpp [hunk 2]
```diff
@@ -148,12 +150,42 @@ PaymentRequest::Show(ErrorResult& aRv)
...
+  }
+
+  RefPtr<PaymentRequestManager> manager = PaymentRequestManager::GetSingleton();
+  nsresult rv;
+  rv = manager->CanMakePayment(mInternalId);
+  if (NS_FAILED(rv)) {
+    promise->MaybeReject(NS_ERROR_FAILURE);
+    return promise.forget();
+  }
+  mResultPromise = promise;
```
Comment (bentian) [inline dom/payments/PaymentRequest.cpp]: nit: newline before this line, to be clearer.

File: dom/payments/PaymentRequest.cpp [hunk 3]
```diff
@@ -148,12 +150,42 @@ PaymentRequest::Show(ErrorResult& aRv)
...
+  ErrorResult result;
+  RefPtr<Promise> promise = Promise::Create(global, result);
+  if (result.Failed()) {
+    aRv.Throw(NS_ERROR_FAILURE);
+    return nullptr;
+  }
+
+  RefPtr<PaymentRequestManager> manager = PaymentRequestManager::GetSingleton();
+  nsresult rv;
+  rv = manager->CanMakePayment(mInternalId);
```
Comment (bentian) [inline dom/payments/PaymentRequest.cpp]: Simplify into 1 line.
`nsresult rv = manager->CanMakePayment(mInternalId);`

File: dom/payments/PaymentRequest.cpp [hunk 4]
```diff
@@ -6,6 +6,7 @@
 
 #include "mozilla/dom/PaymentRequest.h"
 #include "mozilla/dom/PaymentResponse.h"
+#include "PaymentRequestManager.h"
```
Comment (bentian) [inline dom/payments/PaymentRequest.cpp]: nit: alphabetical order.

File: dom/payments/PaymentRequest.cpp [hunk 5]
```diff
@@ -118,10 +119,11 @@ PaymentRequest::Constructor(const GlobalObject& aGlobal,
...
   }
 
-  // TODO : Use PaymentRequestManager::CreatePayment to create PaymentRequest
-  // Need Bug 1345390 to complete the functionality.
-  // We also set mId in "CreatePayment()"
-  return nullptr;
+  RefPtr<PaymentRequest> request;
+  RefPtr<PaymentRequestManager> manager = PaymentRequestManager::GetSingleton();
+  nsresult rv;
+  rv = manager->CreatePayment(window, aMethodData, aDetails, aOptions, getter_AddRefs(request));
```
Comment (bentian) [inline dom/payments/PaymentRequest.cpp]: Simplify `rv` into 1 line, declare `request` close to where it's used, and add some comment as following:
```
RefPtr<PaymentRequestManager> manager = PaymentRequestManager::GetSingleton();

// Create PaymentRequest and set its |mId|
RefPtr<PaymentRequest> request;
nsresult rv = manager->CreatePayment(window, aMethodData, aDetails, aOptions, getter_AddRefs(request));

return NS_FAILED(rv) ? nullptr : request.forget();
```

File: dom/payments/PaymentRequest.cpp [hunk 6]
```diff
@@ -148,12 +150,42 @@ PaymentRequest::Show(ErrorResult& aRv)
 void
 PaymentRequest::RespondCanMakePayment(bool aResult)
 {
+  if (mResultPromise) {
```
Comment (bentian) [inline dom/payments/PaymentRequest.cpp]: Add `MOZ_ASSERT(mResultPromise)` for debugging if `mResultPromise == nulptr` is an abnormal case.
```
MOZ_ASSERT(mResultPromise);

mResultPromise->MaybeResolve(aResult);
mResultPromise = nullptr;
```

**Comment by alphan102** [comment]
@marcoscaceres
In the latest patch, I use "NotAllowedError" and remove the changes in RequestManager.cpp (it will be in the Eden's patch).

**Pull request merged**